src/libos/src/fs/file_ops/sendfile.rs

```diff
@@ -17,33 +17,69 @@ pub fn do_sendfile(
     let out_file = current.file(out_fd)?;
     let mut buffer: [u8; 1024 * 11] = unsafe { MaybeUninit::uninit().assume_init() };
 
-    let mut read_offset = match offset {
+    // starting offset
+    let mut init_offset = match offset {
         Some(offset) => offset,
         None => in_file.seek(SeekFrom::Current(0))?,
     } as usize;
 
-    // read from specified offset and write new offset back
-    let mut bytes_read = 0;
-    while bytes_read < count {
-        let len = min(buffer.len(), count - bytes_read);
-        let read_len = in_file.read_at(read_offset, &mut buffer[..len])?;
-        if read_len == 0 {
-            break;
-        }
-        bytes_read += read_len;
+    // the offset used to read
+    let mut read_offset = init_offset;
+
+    let mut bytes_read_sum: usize = 0;
+    let mut bytes_write_sum: usize = 0;
+
+    'a: while bytes_read_sum < count {
+        let len = min(buffer.len(), count - bytes_read_sum);
+        let read_len = match in_file.read_at(read_offset, &mut buffer[..len]) {
+            Err(e) => {
+                // return error without written bytes
+                if bytes_write_sum == 0 {
+                    return Err(e);
+                }
+                // return error with written bytes
+                break 'a;
+            }
+            Ok(len) => {
+                if len == 0 {
+                    break 'a;
+                }
+                len
+            }
+        };
+
+        bytes_read_sum += read_len;
         read_offset += read_len;
+
+        // the bytes_written means written bytes for every circle
         let mut bytes_written = 0;
-        while bytes_written < read_len {
-            let write_len = out_file.write(&buffer[bytes_written..read_len])?;
+
+        'b: while bytes_written < read_len {
+            let write_len: usize = match out_file.write(&buffer[bytes_written..read_len]) {
+                Err(e) => {
+                    if e.errno() == Errno::EAGAIN {
+                        // return error without written bytes
+                        if bytes_write_sum == 0 {
+                            return Err(e);
+                        }
+                        // some data has been writen
+                        break 'a;
+                    }
+                    return Err(e);
+                }
+                Ok(len) => len,
+            };
+
             if write_len == 0 {
                 return_errno!(EBADF, "sendfile write return 0");
             }
```

Comment:
This error checking logic is not updated by this commit. But since this commit is all about improving error handling, I think it is a good opportunity to put all error checking code under scrutiny.

First of all, I don't expect the write methods of any types of file to return 0 if the length of the write buffer is greater than 0. Second, even when `write_len = 0`, we should still report the success of making some progress when `bytes_written > 0`. Third, the errno number of `EBADF` indicates a file descriptor is not valid, which is not the case (we have indeed opened `out_file`).

I recommend to modify the code this way.

```rust
// A write method should never return Ok(0) if the length of the write buffer is greater than 0.
debug_assert!(write_len > 0);
```

Reply:
Nice suggestion